hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/DataNodeSafeModeRule.java

```diff
@@ -63,16 +92,23 @@ protected boolean validate() {
 
   @Override
   protected void process(NodeRegistrationContainerReport reportsProto) {
+    UUID dnUUID = reportsProto.getDatanodeDetails().getUuid();
+
+    // If a DN is registered for the first time
+    // (as it is possible for the DN to be registered multiple times),
+    // we will write the UUID of this DN into the `registeredDnSet` and
+    // remove it from the unregistered list.
+    if (pipeLineDnSet.contains(dnUUID) || !registeredDnSet.contains(dnUUID)) {
+      registeredDnSet.add(dnUUID);
+      registeredDns = registeredDnSet.size();
+      unRegisteredDn.remove(dnUUID);
```

Comment:
@errose28 Regarding the issue we discussed together in HDDS-11481, I plan to add a variable here to store `unRegisteredDn` and display it in the status. Do you think this approach is acceptable?

Reply:
This is using the the persisted pipeline membership to determine which nodes have not been seen yet? That should work. It won't catch the cases where a new DN not in any pipelines has not registered yet but it at least provides more information.

I'm not sure that putting all nodes back in the unregistered list on refresh it the correct behavior though, since nodes that have already registered should remain accounted for by the rule on refresh.

Reply:
Thank you for your message! I have made adjustments to this part of the logic. During re-registration, only DNs that are not in the `registeredDnSet` will be placed in the `unRegisteredDn`.